### PR TITLE
mgr/pybind/snap_schedule: do not fail when no fs snapshots are available

### PR DESCRIPTION
**Pull request by leseb**
When listing for available snapshot, we should not an error in case
there is none. We should just return 0 with an empty dict.

Fixes: https://tracker.ceph.com/issues/49837
Signed-off-by: Sébastien Han <seb@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/snap_schedule/module.py [hunk 1]
```diff
@@ -92,6 +92,9 @@ def snap_schedule_list(self, path: str,
         except CephfsConnectionException as e:
             return e.to_tuple()
         if not scheds:
+            if format == 'json':
+                out = {}
+                return 0, json.dumps(out), ''
```
Comment (batrick) [inline src/pybind/mgr/snap_schedule/module.py:97]:
````
snap_schedule/__init__.py:9: note: In module imported here:
snap_schedule/module.py: note: In member "snap_schedule_list" of class "Module":
snap_schedule/module.py:96: error: Need type annotation for 'out' (hint: "out: Dict[<type>, <type>] = ...")
````

**Review by batrick — CHANGES_REQUESTED** [review]
Suggest this commit and PR title: `mgr/pybind/snap_schedule: do not fail when no fs snapshots are available`

**Comment by leseb** [comment]
jenkins test make check

**Review by batrick — APPROVED** [review]

**Review by vshankar — COMMENTED** [review]
Loogs good.

Minor nit: test cases needs fix: see qa/tasks/cephfs/test_snap_schedules.py:

- test_non_existent_schedule
- test_snap_schedule_list_post_schedule_remove

File: src/pybind/mgr/snap_schedule/module.py [hunk 2]
```diff
@@ -92,6 +92,9 @@ def snap_schedule_list(self, path: str,
         except CephfsConnectionException as e:
             return e.to_tuple()
         if not scheds:
+            if format == 'json':
+                output: Dict[str, str] = {}
+                return 0, json.dumps(output), ''
```
Comment (jan--f) [inline src/pybind/mgr/snap_schedule/module.py:97]: When there are no schedules this now returns either 0 or an error depending on the `format` option. Is this intentional or should this be consistent?

File: src/pybind/mgr/snap_schedule/module.py [hunk 3]
```diff
@@ -72,7 +72,7 @@ def snap_schedule_get(self,
         except CephfsConnectionException as e:
             return e.to_tuple()
         if format == 'json':
-            json_report = ','.join([ret_sched.report_json() for ret_sched in ret_scheds])
+            json_report = "[" + ','.join([ret_sched.report_json() for ret_sched in ret_scheds]) + "]"
             return 0, f'{json_report}', ''
```
Comment (jan--f) [inline src/pybind/mgr/snap_schedule/module.py:76]:
```suggestion
            json_report = ','.join([ret_sched.report_json() for ret_sched in ret_scheds])
            return 0, f'[{json_report}]', ''
```

**Review by jan--f — CHANGES_REQUESTED** [review]
Just a question and a nit pick. Also PR title and commit should probably be `mgr/pybind/snap_schedule: do not fail when no fs snapshot schedules are available` or similar. This has nothing to do with snapshots being unavailable.

File: src/pybind/mgr/snap_schedule/module.py (hunk 2 above)
Comment (leseb) [reply src/pybind/mgr/snap_schedule/module.py:97]: I was a bit hesitant to apply the same to non-json cases but I felt like `ENOENT` made sense. So it's intentional. The JSON output is meant to be consumed by other tools, if the result is empty there is no need for an error IMO, an empty list is suffient.

**Comment by leseb** [comment]
jenkins test docs

**Comment by leseb** [comment]
> Loogs good.
> 
> Minor nit: test cases needs fix: see qa/tasks/cephfs/test_snap_schedules.py:
> 
> * test_non_existent_schedule
> * test_snap_schedule_list_post_schedule_remove

Done.

**Review by vshankar — APPROVED** [review]

**Comment by vshankar** [comment]
backport note: need to pick PR #41000

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2021-04-23_09:40:03-fs-wip-vshankar-testing-20210423-003830-distro-basic-gibba/

Failure is: https://tracker.ceph.com/issues/50224

**Comment by leseb** [comment]
jenkins test make check

**Comment by leseb** [comment]
jenkins test make check

**Comment by leseb** [comment]
jenkins test make check

**Comment by leseb** [comment]
Looks like the `	160 - mgr-dashboard-frontend-unittests (Failed)` keeps failing, not sure if this is related to my change.

**Comment by vshankar** [comment]
> Looks like the ` 160 - mgr-dashboard-frontend-unittests (Failed)` keeps failing, not sure if this is related to my change.

looks unrelated.

**Comment by vshankar** [comment]
jenkins test make check

**Comment by leseb** [comment]
jenkins test make check

**Comment by leseb** [comment]
jenkins test make check

**Pull request merged**